juspay/hyperswitch
# feat(router): add automatic retries and step up 3ds flow

## Type of Change



## Description
add automatic retries and step up 3ds flow

1. Auto Retries -> When a payment fails due to connector error and if the error is decided to be retryable by Gsm. Then instead of returning error the trackers are updated and the next eligible connector is called based on merchant configuration.

2. Step up 3Ds -> When a no-3ds payment fails due to connector error and if the error is eligible to step up by Gsm. Then instead of returning error the trackers are updated and the payment is retried by enforcing 3ds with same connector based on merchant configuration.

### Additional Changes





## Motivation and Context



## How did you test it?
Tested Manually
1. Enable retries for merchant by enabling `should_call_gsm_{mid}` config
![image](https://github.com/juspay/hyperswitch/assets/56996463/40c6e267-5eb6-4a61-9526-c9d522467fbc)

2. Set max_auto_retries_{mid} config to 1
![image](https://github.com/juspay/hyperswitch/assets/56996463/2d420f2a-92d6-4905-8da1-30e0d12b8f13)

3. Add GSM for relevant connector, code and message to enable retry with step_up_possible as false
<img width="1355" alt="image" src="https://github.com/juspay/hyperswitch/assets/56996463/68ea2f30-0e5c-4ec9-bb31-4e169f7ee3e2">

4. Make a payment to replicate the payment failure with code and message configured in Gsm, can be seen that payment is retried with next eligible connector (attempt_count = 2 and can see detailed view using payments retrieve api with expand_attempts = true flag)
![image](https://github.com/juspay/hyperswitch/assets/56996463/a94991b5-eecb-481b-aea3-bb0374758bf3)
![image](https://github.com/juspay/hyperswitch/assets/56996463/7d656e04-bc40-4431-9385-366c159358f2)

5. Now enable step_up_enabled_{mid} config for merchant
![image](https://github.com/juspay/hyperswitch/assets/56996463/2d83d0f7-a7ec-4a90-9312-eae96b40ec3e)

6. Update Gsm to enable step_up_possible
![image](https://github.com/juspay/hyperswitch/assets/56996463/5850ec45-405f-4cef-8d47-deb4a59619e0)

7. Now, redo the same payment can be seen that instead of retrying with next connector it gets retried with same connector but now enforcing three_ds
![image](https://github.com/juspay/hyperswitch/assets/56996463/77e10959-af8e-41d8-88b0-0c0b64099d71)
![image](https://github.com/juspay/hyperswitch/assets/56996463/8cf0e2e2-847e-4b1a-96ce-13cc09b84965)

## Checklist
